src/cogent3/app/data_store_new.py

```diff
@@ -0,0 +1,340 @@
+_MD5_TABLE = "md5"
+
+
+class IfExist(Enum):
+    SKIP = "skip"
+    OVERWRITE = "overwrite"
+    RAISE = "raise"
+    READONLY = "readonly"
+
+
+SKIP = IfExist.SKIP
+OVERWRITE = IfExist.OVERWRITE
+RAISE = IfExist.RAISE
+READONLY = IfExist.READONLY
+
+
+class DataMemberABC(ABC):
+    def __init__(self, data_store: "DataStoreABC" = None):
+        self.data_store = data_store
+
+    @property
+    @abstractmethod
+    def unique_id(self):
+        ...
+
+    def read(self) -> str:
+        return self.data_store.read(self.unique_id)
+
+    def write(self, data: str) -> None:
+        self.data_store.write(self.unique_id, data)
+
+
+class DataMember(DataMemberABC):
+    def __init__(self, data_store: "DataStoreDirectory" = None, unique_id: str = None):
+        super().__init__(data_store)
+        self._unique_id = unique_id
+
+    def __repr__(self):
+        return self._unique_id
+
+    @property
+    def unique_id(self):
+        return self._unique_id
+
+
+class DataStoreABC(ABC):
+    def __init__(
+        self,
+        source,
+        if_dest_exists: IfExist = READONLY,
+        if_member_exists: IfExist = RAISE,
+    ):
+        self._if_dest_exists = if_dest_exists
+        self._if_member_exists = if_member_exists
+
+    @abstractmethod
+    def __repr__(self):
+        ...
+
+    def __getitem__(self, index):
+        return self.members[index]
+
+    def __len__(self):
+        return len(self.members)
+
+    def __contains__(self, identifier):
+        """whether relative identifier has been stored"""
+        return any(identifier == m.unique_id for m in self)
+
+    @abstractmethod
+    def read(self, unique_id: str) -> str:
+        ...
+
+    @abstractmethod
+    def write(self, unique_id: str, data: str) -> None:
+        if self._if_dest_exists is READONLY:
+            raise IOError("datastore is readonly")
+        if self._if_member_exists is RAISE and unique_id in self:
+            raise IOError("member exists")
+
+    @abstractmethod
+    def write_incomplete(self, unique_id: str, data):
+        if self.if_member_exists is RAISE and unique_id in self:
+            raise IOError("member exists")
+
+    @abstractmethod
+    def write_log(self, unique_id: str, data):
+        if self.if_member_exists is RAISE and unique_id in self:
+            raise IOError("member exists")
+
+    @abstractmethod
+    def describe(self):
+        ...
+
+    @property
+    @abstractmethod
+    def members(self):
+        ...
+
+    @abstractmethod
+    def open(self, member):
+        ...
+
+    def __iter__(self):
+        yield from self.members
+
+    @property
+    @abstractmethod
+    def logs(self):
+        ...
+
+    @property
+    @abstractmethod
+    def incomplete(self):
+        ...
+
+    @property
+    @abstractmethod
+    def summary_logs(self):
+        ...
+
+    @property
+    @abstractmethod
+    def summary_incomplete(self):
+        ...
+
+
+class DataStoreDirectory(DataStoreABC):
+    store_suffix = None
```

Comment:
this is an instance attribute since defined in `__init__` so remove from here (plus we have no plan to subclass like I did in first implementation)